interface.go

```diff
@@ -20,6 +21,11 @@ type LegPublisher interface {
 
 // LegSubscriber is an interface for watching a published dag.
 type LegSubscriber interface {
+	Subscribe(context.Context, SubFnHandler) error
 	OnChange() (chan cid.Cid, context.CancelFunc)
 	Close(context.Context) error
 }
+
+// SubFnHandler is the function handler to run every time a root is updated
```

Comment:
it seems that this replaces the `OnChange` stream of head updates. Consider removing that?

Reply:
can we change `SubFnHandler` to be a function to decide if a `*pubsub.Message` should be considered for a legs update? so something like

```
SetPolicyHandler(PolicyHandler) error
type PolicyHandler func(context.Context, *pubsub.Message) (allowProcessing bool, err error)
```

Reply:
That actually makes even more sense and it addresses what I was trying to achieve with `SubFnHandler`. Will include it right away. Thanks!

Reply:
I just realized we may not be able to use `go-legs` for the reference provider if we use a single pubsub channel to broadcast the advertisement of every provider (as long as we don't add additional data in the pubsub message), as indexers we'll need to inspect Advertisements to check if it belongs to a provider they are indexing:
- Provider publishes Advertisement CID in common PubSub channel.
- Indexer sees CID for a new advertisement and starts a data transfer to fetch just the advertisement.
- Indexer checks if advertisement is from a provider it is indexing.
- If yes, indexer fetches the full DAG recursively and stops in the last advertisement seen for that provider.

I don't think we can embed this logic in a `SetPolicyHandler` or a `SubFnHandler`, we'll need to implement an adaptation of `go-legs` for the provider. What do you think?

Reply:
won't the indexer know if it's worth pulling based on the libp2p identity of the provider that it connects to in step 2?

Reply:
Oh wait! I think you are right. Do we have a way to check the peer that originally published a pubsub message to a topic from a`pubsub.Message`? I thought the `msg.From` of `*pubsub.Message` specified who you were receiving the message from, not the one that generated it. If we have a way to know from the message who published then forget what I said 🤦‍♂️ (and it seems so because I see there's a `GetFrom()` and a `ReceivedFrom()` in the pubsub code).

Reply:
https://github.com/libp2p/go-libp2p-pubsub/blob/master/pubsub.go#L218 ?